src/providers/trusted_service/context/mod.rs

```diff
@@ -0,0 +1,290 @@
+    non_camel_case_types,
+    non_upper_case_globals,
+    clippy::unseparated_literal_suffix,
+    // There is an issue where long double become u128 in extern blocks. Check this issue:
+    // https://github.com/rust-lang/rust-bindgen/issues/1549
+    improper_ctypes,
+    missing_debug_implementations,
+    trivial_casts,
+    clippy::all,
+    unused,
+    unused_qualifications
+)]
+pub mod ts_binding {
+    include!(concat!(env!("OUT_DIR"), "/ts_bindings.rs"));
+}
+
+mod asym_sign;
+pub mod error;
+mod key_management;
+
+#[allow(
+    non_snake_case,
+    non_camel_case_types,
+    non_upper_case_globals,
+    clippy::unseparated_literal_suffix,
+    // There is an issue where long double become u128 in extern blocks. Check this issue:
+    // https://github.com/rust-lang/rust-bindgen/issues/1549
+    improper_ctypes,
+    missing_debug_implementations,
+    trivial_casts,
+    clippy::all,
+    unused,
+    unused_qualifications
+)]
+mod ts_protobuf {
+    include!(concat!(env!("OUT_DIR"), "/ts_crypto.rs"));
+
+    /// Trait for associating an Opcode with each operation type
+    /// and obtaining it in a generic way.
+    pub trait GetOpcode {
+        fn opcode(&self) -> Opcode;
+    }
+
+    macro_rules! opcode_impl {
+        ($type:ty, $opcode:ident) => {
+            impl GetOpcode for $type {
+                fn opcode(&self) -> Opcode {
+                    Opcode::$opcode
+                }
+            }
+        };
+
+        ($type_in:ty, $type_out:ty, $opcode:ident) => {
+            impl GetOpcode for $type_in {
+                fn opcode(&self) -> Opcode {
+                    Opcode::$opcode
+                }
+            }
+
+            impl GetOpcode for $type_out {
+                fn opcode(&self) -> Opcode {
+                    Opcode::$opcode
+                }
+            }
+        };
+    }
+
+    opcode_impl!(OpenKeyIn, OpenKeyOut, OpenKey);
+    opcode_impl!(CloseKeyIn, CloseKey);
+    opcode_impl!(GenerateKeyIn, GenerateKeyOut, GenerateKey);
+    opcode_impl!(DestroyKeyIn, DestroyKeyOut, DestroyKey);
+    opcode_impl!(SignHashIn, SignHashOut, SignHash);
+    opcode_impl!(VerifyHashIn, VerifyHashOut, VerifyHash);
+    opcode_impl!(ImportKeyIn, ImportKeyOut, ImportKey);
+    opcode_impl!(ExportPublicKeyIn, ExportPublicKeyOut, ExportPublicKey);
+
+    /// Trait allowing the handle of opened-key-dependent operations
+    /// to be set in a generic way.
+    pub trait SetHandle {
+        fn set_handle(&mut self, handle: u32);
+    }
+
+    macro_rules! set_handle_impl {
+        ($type:ty) => {
+            impl SetHandle for $type {
+                fn set_handle(&mut self, handle: u32) {
+                    self.handle = handle;
+                }
+            }
+        };
+    }
+
+    set_handle_impl!(DestroyKeyIn);
+    set_handle_impl!(SignHashIn);
+    set_handle_impl!(VerifyHashIn);
+    set_handle_impl!(ExportPublicKeyIn);
+}
+
+// TODO:
+// * RPC caller error handling
+// * proper logging
```

Comment:
Would be nice to create issues for those if still needed as well

Reply:
Logging has been covered, it's just error handling now and I've raise #332 for that